README
Perl interface to the PC/SC smart card library
==============================================

This package contains a Perl wrapper to the PC/SC smartcard library
(pcsc-lite) [1] from MUSCLE [2] together with some small examples.

You can find applications using this wrapper in the pcsc-tools package
[3].

[1] http://pcsclite.alioth.debian.org/
[2] http://www.musclecard.com/
[3] http://ludovic.rousseau.free.fr/softwares/pcsc-tools/


Build and Install:
==================

A very quick documentation on how to build and install the wrapper.

$ perl Makefile.PL
$ make

Insert a smart card in the (first) PCSC reader

$ make test
$ sudo make install


Remarks:
========

For Windows users:

Status behaves different on Unix and from Win32
	it returns the Atr len
	it does not return any error like


Authors:
========

- Lionel VICTOR for the main part of the job
- Ludovic ROUSSEAU for the debug and maintainance


Licences:
=========

    pcsc-perl: wrapper to the PC/SC smartcard library
    Copyright (C) 2001-2003  Lionel VICTOR
    Copyright (C) 2001-2010  Ludovic ROUSSEAU

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


History:
========

1.4.16 - 20 November 2023, Ludovic ROUSSEAU
    - provide a META.json file to comply with CPAN::Meta::Spec

1.4.15 - 18 November 2023, Ludovic ROUSSEAU
    - add support of SCARD_W_SECURITY_VIOLATION error code

1.4.14 - 19 November 2015, Ludovic ROUSSEAU
    - Fix _StringifyError() on Mac OS X. The string version of
      PCSC::errno now works.

1.4.13 - 1 April 2013, Ludovic ROUSSEAU
    - Fix a warning with Perl 5.16

1.4.12 - 6 March 2011, Ludovic ROUSSEAU
    - Also check for SVt_PVIV and not just SVt_IV.
    Fixes Debian bug #613722 "libpcsc-perl: GetStatusChange error after print"

1.4.11 - 27 October 2010, Ludovic ROUSSEAU
    - Do not use SCardSetTimout() since it is no more present in
      pcsc-lite > 1.6.4

1.4.10 - 18 August 2010, Ludovic ROUSSEAU
    - Use LPSCARD_READERSTATE instead of LPSCARD_READERSTATE_A since it
      is no more defined in pcsc-lite >= 1.6.2

1.4.9 - 30 June 2010, Ludovic ROUSSEAU
    - remove reference to SCARD_W_INSERTED_CARD since it is no more
      define in pcsc-lite 1.6.0 Thanks to Olivier Huber for the bug
      report

1.4.8 - 23 September 2009, Ludovic ROUSSEAU
    - minor bugs fixed

1.4.7 - 26 March 2008, Ludovic ROUSSEAU
    - merge all the Makefile.PL.* in Makefile.PL

1.4.6 - 7 March 2007, Ludovic ROUSSEAU
    - add support of GNU/kFreeBSD

1.4.5 - 2 March 2007, Ludovic ROUSSEAU
    - remove some pcsc-lite specific constants to make it compile
      against pcsc-lite 1.4.0

1.4.4 - 12 August 2006, Ludovic ROUSSEAU
    - add support of extended APDU

1.4.3 - 16 May 2006, Ludovic ROUSSEAU
    - update for Mac OS X Tiger

1.4.2 - 6 August 2004, Ludovic ROUSSEAU
    - allow to use SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1 as the protocol
      this is used by gscriptor 1.4.0

1.4.1 - 4 July 2004, Ludovic ROUSSEAU
    - Card/Card.pm:
     . ISO7816Error(): return an error if given $sw if not 5 chars long
       "xx xx"
     . Transmit(): return (undef) if the protocol is not defined. This
       happens if you change a T=0 card with a T=1 (or vice versa) and
       try to exchange an APDU without disconnecting first.
    - PCSCperl.h: replace LPCSTR by LPCTSTR (pcsc-lite 1.2.9 beta4)

1.4.0 - 20 May 2004, Ludovic ROUSSEAU
    - use a new Control() API to reflect the change in SCardControl() in
      pcsc-lite. We now have
      $RecvData = $hCard->Control($control_code, \@SendData);
    - load libpcsclite.so.1 instead of libpcsclite.so.0

1.3.1 - 2 April 2004, Ludovic ROUSSEAU
    - do not use PCSCLITE_MAX_CHANNELS anymor since it is no more
      defined by pcsc-lite > 1.2.0 and is not defined under Windows
    - test.pl: add demo code for GetStatusChange()
    - PCSC.pod: document GetStatusChange()
    - allow commands to be in the form "00A40100020100" instead of just
      "00 A4 01 0 0 02 01 00"

1.3.0 - 13 December 2003, Ludovic ROUSSEAU
    - add support of SCardControl with $out = $hContext->Control (\@in);
      Thanks to Andrew Kay.
    - more support of Windows PC/SC and compiler.
      Thanks to Andrew Kay and Troy Curtiss
    - array_to_ascii() returns an empty string for an empty list
    - add Makefile_win.PL for Windows
    - Card.pm: by default use T=0 | T=1 protocols instead of just T=0

1.2.2 - 27 May 2003, Ludovic ROUSSEAU
    - add two constant definitions in PCSC.pm. Thanks to Ville Skyttä
      for report and patch
    - add a PL_FILES section in Makefile*.PL to avoid problem since we
      have Makefile.PL and Makefile_OSX.PL. Again thanks goes to Ville
      Skyttä
    - Also thanks to Wolfgang Hommel for uploading the package to CPAN
      and registering the official name Chipcard::PCSC (Wolfgang is
      maintainer of the Chipcard::CTAPI wrapper)

1.2.1 - 24 May 2003, Ludovic ROUSSEAU
    - redesign the Makefile.PL to make the modules at least installable
    - add Card/Makefile.PL to also install Chipcard::PCSC::Card
    - modified test.pl to accept any SW in TransmitWithCheck

1.2.0 - 9 May 2003, Ludovic ROUSSEAU
    - naming scheme migration from PCSC to Chipcard::PCSC
    - add TransmitWithCheck() and Chipcard::PCSC::Card::ISO7816Error()
      functions

1.1.3 - 7 Nov 2002, Ludovic ROUSSEAU
    - small modifications to support Perl 5.8

1.1.2 - 28 Aug 2002, Ludovic ROUSSEAU
    - small modification to link against libpcsclite.so.0 instead of
      libpcsclite.so since libpcsclite.so should only be available on
      development platforms (-dev packages).

1.1.1 - 7 Mar 2002, Ludovic ROUSSEAU
    - small typesetting modifications od .pod files

1.1.0 - 6 Mar 2002, Ludovic ROUSSEAU
    - create this README file with the history
    - complete reindentation of .pod files

1.0.9 - 5 Mar 2002, Ludovic ROUSSEAU
    - the minium APDU length is 4 bytes and not 5
    - correctly test if a digit is hexa or not

1.0.8 - 18 Oct 2001, Ludovic ROUSSEAU



Changelog
2013-04-01  author  <author@debian>

	* create_distrib.sh:
	rcs2log has been moved in /usr/share/cvs/contrib/rcs2log

	* PCSC.pm: Support Perl 5.16

	Remove the warning:
	defined(@array) is deprecated at /usr/lib/perl/5.16.3/Chipcard/PCSC.pm line 69.
		(Maybe you should just omit the defined()?)

	Thanks to Viliam Pucik for the bug report

2011-03-06  author  <author@debian>

	* README: release 1.4.12

	* PCSC.xs: Also check for SVt_PVIV and not just SVt_IV.

	Fixes Debian bug #613722 "libpcsc-perl: GetStatusChange error after print"

	* PCSC.xs, PCSC.pm, test.pl: Use lines of less than 80 columns

2010-10-27  author  <author@debian>

	* create_distrib.sh: compress using bzip2 instead of gzip

	* README: release 1.4.11

	* PCSCperl.h: SCardCancelTransaction() is no more present in pcsc-lite

	* PCSC.pm, PCSC.pod, PCSC.xs, PCSCperl.h:
	SCardSetTimout is no more present in pcsc-lite > 1.6.4
	It was a pcsc-lite specific and deprecated function

2010-08-18  author  <author@debian>

	* PCSCperl.h:
	use SCARD_READERSTATE * instead of LPSCARD_READERSTATE since is not
	define in pcsc-lite < 1.6.3

	* README: release 1.4.10

	* PCSC.pm, PCSC.xs, PCSCperl.h, README: Update copyright date

	* create_distrib.sh: Allow more than one digit in the version numbers

	* PCSC.xs, PCSCperl.h:
	Use LPSCARD_READERSTATE instead of LPSCARD_READERSTATE_A since it is no
	more defined in pcsc-lite >= 1.6.2

2010-06-30  author  <author@debian>

	* README: release 1.4.9

	* PCSC.pm, PCSC.pod, PCSC.xs: sort constants in alphabetical order

	* PCSC.pod, PCSC.xs:
	SCARD_E_UNSUPPORTED_FEATURE is not specific to pcsc-lite but is also
	available on Windows

	* PCSC.pm, PCSC.pod, PCSC.xs:
	SCARD_W_INSERTED_CARD is no more defined in pcsc-lite 1.6.0 and then
	pcsc-perl failed to build. It was an error code specific to pcsc-lite.

2010-01-03  author  <author@debian>

	* Card/Card.pod: fix spelling error

2009-09-23  author  <author@debian>

	* README: release 1.4.8

	* Card/Card.pod: Correct 3 bugs signaled by podchecker

2009-09-06  author  <author@debian>

	* PCSC.xs:
	PCSC.xs:853: warning: format ‘%d’ expects type ‘int’, but argument 2 has type ‘long unsigned int’

2008-09-28  author  <author@debian>

	* PCSC.pod: example code for GetStatusChange()

2008-03-26  author  <author@debian>

	* Card/Card.pm: type: prefered -> preferred

	* Card/Card.pm: update copyright date

	* Card/Card.pm: typo: prefered -> preferred

	* Card/Card.pod: typos

	* README: release 1.4.7

2008-03-12  author  <author@debian>

	* MANIFEST: remove removed files (merged)

	* Makefile_OSX.PL, Makefile_win.PL: merged in Makefile.PL

	* README, README.OSX, README.Unix, README.Windows:
	merge all README.* in README

	* Makefile.PL: merge Makefile_win.PL and Makefile_OSX.PL

	* PCSC.pod: typos

	* PCSC.pod: typo

	* PCSC.pm: version 0.05

	* PCSC.pod, PCSCperl.h: update copyright date

	* PCSCperl.h:
	reorder the .h inclusion to have a default for Unix system

2007-03-07  author  <author@debian>

	* README: release 1.4.6

	* PCSCperl.h: add support of GNU/FreeBSD

	Thanks to Cyril Brulebois for the patch (Debian bug #413618)

2007-03-02  author  <author@debian>

	* README: release 1.4.5

	* MANIFEST: add Card/t/test.t

	* PCSC.xs:
	remove a const on the return type to avoid: PCSC.xs:244: warning: type
	qualifiers ignored on function return type

	* PCSCperl.h: For Linux
	#define LPCTSTR LPCSTR
	#define LPTSTR LPSTR
	to avoid using deprecated types

2007-02-05  author  <author@debian>

	* PCSC.xs: remove some pcsc-lite specific constants

2006-12-09  author  <author@debian>

	* t/test.t:
	remove Chipcard::PCSC::Card tests since they are now done by Card/t/test.t

	* Card/t/test.t: new file. Copy from ../t/test.t

	* t/test.t, test.pl: use Test::More instead of ExtUtils::testlib

2006-10-10  author  <author@debian>

	* PCSC.pm: remove a debug print command

2006-10-08  author  <author@debian>

	* PCSC.pm: version 0.04

	* PCSC.pm:
	ascii_to_array(): check that an hex number is exactly two characters

2006-08-12  author  <author@debian>

	* README: release 1.4.4

	* Card/Card.pm, examples/gsm_directory.pl, examples/test_iso_error.pl, test/multiple_readers.pl, test/single_reader.pl, PCSC.pm, README.Unix, README.Windows, create_distrib.sh, test.pl:
	new $Id$ format

2006-05-30  author  <author@debian>

	* PCSC.xs, PCSCperl.h: add support of extended APDU

2006-05-16  author  <author@debian>

	* README: release 1.4.3

2006-05-04  author  <author@debian>

	* PCSC.xs: use unsigned char for pbAtr[] to have "3B FA ..." instead of
	"3B FFFFFFFA ..."

	* t/test.t: update $Id$

	* Makefile_OSX.PL: update for Mac OS X Tiger

	* PCSC.xs:
	change some types and use some cast to make the code compile under Mac
	OS X

	* PCSCperl.h: add some #include <> for Mac OS X

2004-08-06  author  <author@debian>

	* README: release 1.4.2

	* PCSC.xs:
	allow to use SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1 as the protocol

	* Card/Card.pm: typo: informatin -> information

2004-07-04  author  <author@debian>

	* README: release 1.4.1

	* README:
	use http://pcsclite.alioth.debian.org/ as the pcsc-lite homepage

	* Makefile.PL, PCSCperl.h:
	back to use pcsclite.h instead of PCSC/pcsclite.h for __linux__ but now
	use `pkg-config --cflags libpcsclite` at compilation

2004-06-29  author  <author@debian>

	* Card/Card.pm:
	ISO7816Error(): return an error if given $sw if not 5 chars long "xx xx"

	* Card/Card.pm:
	return (undef) if the protocol is not defined in Transmit. This happens
	if you change a T=0 card with a T=1 (or vice versa) and try to exchange
	an APDU without disconnecting first.

2004-06-25  author  <author@debian>

	* Card/Card.pod: add en empty line

	* PCSCperl.h: - use PCSC/pcsclite.h
	- replace LPCSTR by LPCTSTR

2004-05-30  author  <author@debian>

	* README: release 1.4.0

	* PCSC.pm: ABI changed so change VERSION number

	* PCSCperl.h: - add support of new SCardControl() API through Control()
	- now loads libpcsclite.so.1 instead of libpcsclite.so.0


examples/gsm_directory.pl
#!/usr/bin/perl


#    This program reads the phone directory of a GSM11.11 SIM card and
#    prints out its contents in a human readable format. It is based on
#    another Perl script from Ludovic ROUSSEAU <ludovic.rousseau@free.fr>
#    That you can obtain at the following URL:
#        http://ludovic.rousseau.free.fr/softwares/SIM-1.0.tar.gz
#
#    Copyright (C) 2001 - Lionel VICTOR, 2003 Ludovic ROUSSEAU
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA


# $Id: gsm_directory.pl,v 1.6 2003/05/06 21:23:30 rousseau Exp $


use warnings;
use strict;
use Carp;

use Chipcard::PCSC;
use Chipcard::PCSC::Card;

use Getopt::Std;
use Term::ReadKey;

my %options;

my $hContext = new Chipcard::PCSC();
my $hCard;
my $RecvData;
my $nRecordsLength;
my $PIN;

die ("Could not create PCSC object: $Chipcard::PCSC::errno\n") unless defined $hContext;

sub gsmrecord_to_ascii {
	my $gsmrecord = shift;
	my $out_string = '';
	my $delta = $#$gsmrecord-27;
	my $i;

	if ($$gsmrecord[0] != 0xFF) {
		for $i (0..14+$delta-1) {
			if ($$gsmrecord[$i] != 0xFF) {
				$out_string.=chr($$gsmrecord[$i])
			} else {
				$out_string.=' ';
			}
		}

		my $phone_length = $$gsmrecord[14] - 2;
		if ($phone_length > 0) {
			my $digit;
			$out_string .= '-> ';
			for $i (16+$delta..$#$gsmrecord) {
				$digit = $$gsmrecord[$i]&0x0F;
				$out_string .= sprintf ("%01X", $digit) if ($digit!=0x0F);
				$digit = ($$gsmrecord[$i]&0xF0)>>4;
				$out_string .= sprintf ("%01X", $digit) if ($digit!=0x0F);
			}
		} else {
			$out_string.= "null or invalid phone";
		}
	}

	return $out_string;
}

sub pin_to_array {
	my $pin = shift;
	my @array;

	confess ("PIN code must not exceed 8 bytes.") unless (length $pin <= 8);

	$pin =~ s/(.)/$1 /g;
	@array = split / /, $pin;
	@array = map (ord, @array);

	# Pad the array with 0xFF up to 8 bytes
	while ($#array < 7) { push @array, 0xFF; }

	return @array;
}

getopt ('hvr:', \%options);

if (exists $options{h}) {
	#TODO
	print "TODO: usage -v verbose -r reader -h help";
}

if (exists $options{r}) {
    $hCard = new Chipcard::PCSC::Card ($hContext, $options{r});
    die ("Can't allocate PCSCCard object: $Chipcard::PCSC::errno\n") unless defined $hCard;
    print STDERR "Using given card reader: $options{r}\n" if exists $options{'v'};
} else {
    my @readers_list = $hContext->ListReaders ();
    die ("Can't get readers list: $Chipcard::PCSC::errno\n") unless defined $readers_list[0];
    print STDERR "No reader given: using $readers_list[0]\n" if exists $options{v};
    $hCard = new Chipcard::PCSC::Card ($hContext, $readers_list[0]);
    die ("Can't allocate PCSCCard object: $Chipcard::PCSC::errno\n") unless defined $hCard;
}

# Select MF (3F00)
print STDERR "Selecting Master File (3F00)\n" if exists $options{v};
$RecvData = $hCard->Transmit([0xA0, 0xA4, 0x00, 0x00, 0x02, 0x3F, 0x00]);
die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;
die ("Can't select MF: SW = [ ".Chipcard::PCSC::array_to_ascii($RecvData)." ]") unless $$RecvData[0] == 0x9F;

# Select DF Telecom (7F10)
print STDERR "Selecting Telecom Directory (3F00/7F10)\n" if exists $options{v};
$RecvData = $hCard->Transmit([0xA0, 0xA4, 0x00, 0x00, 0x02, 0x7F, 0x10]);
die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;
die ("Can't select DF Telecom: SW = [ ".Chipcard::PCSC::array_to_ascii($RecvData)." ]") unless $$RecvData[0] == 0x9F; 

# Select EF_ADN (6F3A) (Abbreviated Dialing Numbers)
print STDERR "Selecting Abbreviated Dialing Numbers File (3F00/7F10/6F3A)\n" if exists $options{v};
$RecvData = $hCard->Transmit([0xA0, 0xA4, 0x00, 0x00, 0x02, 0x6F, 0x3A]);
die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;
die ("Can't select EF_ADN: SW = [ ".Chipcard::PCSC::array_to_ascii($RecvData)." ]") unless $$RecvData[0] == 0x9F;

# Get Response (get informations about EF_ADN)
# The last SW gives the length of available bytes:
# 9F xx means there are xx bytes waiting
print STDERR "Getting ADN Records length\n" if exists $options{v};
$nRecordsLength = $$RecvData[1];
$RecvData = $hCard->Transmit([0xA0, 0xC0, 0x00, 0x00, $nRecordsLength]);
die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;
die ("Can't retrieve records length: SW = [ ".Chipcard::PCSC::array_to_ascii($RecvData)." ]") unless $$RecvData[$nRecordsLength] == 0x90;

# Extract record length from the response
$nRecordsLength = $$RecvData[$nRecordsLength-1];
print STDERR "Records are $nRecordsLength bytes long\n" if exists $options{v};

# Asks the user for his/her PIN code
print STDOUT "Please insert your PIN code:";
ReadMode 'noecho';
$PIN = ReadLine 0;
ReadMode 'normal';
chomp $PIN;
print STDOUT "\n";

# Submitting PIN code to the card
print STDERR "Submitting PIN code\n" if exists $options{v};
$RecvData = $hCard->Transmit([0xA0, 0x20, 0x00, 0x01, 0x08, pin_to_array($PIN)]);
die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;
die ("Can't access file (wrong PIN code ?): SW = [ ".Chipcard::PCSC::array_to_ascii($RecvData)." ]") unless $$RecvData[0] == 0x90;

# Get phone records
my $i;
my @SW;
print STDERR "Getting records\n" if exists $options{v};
for $i (1..255) {
	$RecvData = $hCard->Transmit([0xA0, 0xB2, $i, 0x04, $nRecordsLength]);
	die ("Can't communicate: $Chipcard::PCSC::errno\n") unless defined $RecvData;

	# pop the status word out of the retrieved bytes
	$SW[1] = pop @$RecvData;
	$SW[0] = pop @$RecvData;

	# Ignore referencing errors (record out of range, file
	# empty, etc...) but die if anything else bad occurs.
	die ("Can't read from EF_ADN: SW = [ ".sprintf ("%02X %02X",$SW[0],$SW[1])." ]\n") if ($SW[0] != 0x94 && $SW[0] != 0x90);

	# Exit if the status word is not 90xx
	last if $SW[0] != 0x90;

	# TODO:remove that it is only debug...
	# TODO: format that and print phone record...
	print "SLOT $i: ".gsmrecord_to_ascii($RecvData)."\n";
	#print "[ ".Chipcard::PCSC::array_to_ascii($RecvData)." ] [".Chipcard::PCSC::array_to_ascii(\@SW)."]\n";
}

print STDERR "Reseting the card\n" if exists $options{v};
$hCard->Disconnect($Chipcard::PCSC::SCARD_RESET);

# End of File

